src/ui/Frame.cpp

```diff
@@ -0,0 +1,1164 @@
+			Button* button = *it;
+			prev = std::next(it);
+
+			if (!destWidget) {
+				destWidget = button->handleInput();
+			}
+
+			Button::result_t buttonResult = button->process(_size, actualSize, usable);
+			if (usable && buttonResult.highlighted) {
+				result.highlightTime = buttonResult.highlightTime;
+				result.tooltip = buttonResult.tooltip;
+				if (buttonResult.clicked) {
+					button->activate();
+				}
+				result.usable = usable = false;
+			}
+
+			if (destWidget && button->isSelected()) {
+				button->deselect();
+			}
+		}
+	}
+
+	// process (widget) sliders
+	{
+		auto prev = sliders.rbegin();
+		for (auto it = sliders.rbegin(); it != sliders.rend(); it = prev) {
+			Slider* slider = *it;
+			prev = std::next(it);
+
+			if (!destWidget && !slider->isActivated()) {
+				destWidget = slider->handleInput();
+			} else {
+				slider->control();
+			}
+
+			Slider::result_t sliderResult = slider->process(_size, actualSize, usable);
+			if (usable && sliderResult.highlighted) {
+				result.highlightTime = sliderResult.highlightTime;
+				result.tooltip = sliderResult.tooltip;
+				if (sliderResult.clicked) {
+					slider->fireCallback();
+				}
+				result.usable = usable = false;
+			}
+
+			if (destWidget && slider->isSelected()) {
+				slider->deselect();
+			}
+		}
+	}
+
+	// process the frame's list entries
+	if (usable && list.size() > 0) {
+		int num = 0;
+		entry_t* prev = nullptr;
+		auto next = list.begin();
+		for (auto it = list.begin(); it != list.end(); it = next) {
+			next = std::next(it);
+
+			entry_t* entry = *it;
+			if (entry->suicide) {
+				if (selection == it) {
+					selection = selection == list.begin() ? list.begin() : --selection;
+				}
+				delete entry;
+				list.erase(it);
+				continue;
+			}
+
+			SDL_Rect entryRect;
+			entryRect.x = _size.x + border - actualSize.x; entryRect.w = _size.w - border * 2;
+			entryRect.y = _size.y + border + num * entrySize - actualSize.y; entryRect.h = entrySize;
+
+			if (rectContainsPoint(_size, omousex, omousey) && rectContainsPoint(entryRect, omousex, omousey)) {
+				result.highlightTime = entry->highlightTime;
+				result.tooltip = entry->tooltip.c_str();
+				if (mousestatus[SDL_BUTTON_LEFT]) {
+					if (!entry->pressed) {
+						entry->pressed = true;
+						activateEntry(*entry);
+					}
+				} else {
+					entry->pressed = false;
+					Widget::Args args(entry->params);
+					if (entry->highlighting) {
+						(*entry->highlighting)(args);
+					}
+					if (!entry->highlighted) {
+						entry->highlighted = true;
+						if (entry->highlight) {
+							(*entry->highlight)(args);
+						}
+					}
+				}
+				result.usable = usable = false;
+			} else {
+				entry->highlightTime = SDL_GetTicks();
+				entry->highlighted = false;
+				entry->pressed = false;
+			}
+
+			++num;
+			prev = entry;
+		}
+	}
+
+	// process fields
+	{
+		auto prev = fields.rbegin();
+		for (auto it = fields.rbegin(); it != fields.rend(); ++it) {
```

Comment:
Why do these 3 for loops look different?

	// process (widget) sliders
		for (auto it = sliders.rbegin(); it != sliders.rend(); **it = prev**) {

	// process the frame's list entries
		for (auto it = list.begin(); it != list.end(); **it = next**) {

	// process fields
		for (auto it = fields.rbegin(); it != fields.rend(); **++it**) {

Reply:
most ui elements are drawn one way and processed in reverse, so the last inserted element is most visible and also captures inputs first. this is not important for a frame list so it processes beginning to end.

Reply:
Ya, I missed the rbegin/begin difference on my read-through, so I understand why the second looks different from the other two...

I guess my next question of why do 1 and 3 look different, would be answered by:
> I don't embed it = std::next(it) in the for loop because the iterator could be invalid after the process runs (for example, a button that deletes itself). the next position is therefore saved so the increment is valid

I'm assuming fields don't delete themselves (but sliders do), so we don't need to save the incrementor.